lib/configuration.js

```diff
@@ -50,201 +53,274 @@ function applyLoggingOptions(options) {
-  outConfig.options.reporter = coerceToArray(outConfig.options.reporter);
-  outConfig.options.output = coerceToArray(outConfig.options.output);
-  outConfig.options.header = coerceToArray(outConfig.options.header);
-  outConfig.options.method = coerceToArray(outConfig.options.method);
-  outConfig.options.only = coerceToArray(outConfig.options.only);
-  outConfig.options.path = coerceToArray(outConfig.options.path);
-  outConfig.options.hookfiles = coerceToArray(outConfig.options.hookfiles);
-
-  outConfig.options.method = outConfig.options.method
-    .map(method => method.toUpperCase());
-
-  if (outConfig.options.user) {
-    const token = Buffer.from(outConfig.options.user).toString('base64');
-    outConfig.options.header.push(`Authorization: Basic ${token}`);
-    delete outConfig.options.user;
+/**
+ * Returns the errors and warnings relative to the given config.
+ */
+const validateConfig = config => ({
+  warnings: flushMessages(deprecatedOptions, config),
+  errors: flushMessages(unsupportedOptions, config),
+});
+
+const normalizeConfig = R.compose(
+  coerceOptions,
+  removeUnsupportedOptions
+);
+
+/**
+ * Flattens giving configuration Object, removing nested "options" key.
+ * This makes it possible to use nested "options" key without introducing
+ * a breaking change to the library's API.
+ * @todo Remove this method after "options" key is deprecated in the next
+ * major version. Document the removal of "options" in the Dredd docs.
+ */
+function flattenConfig(config) {
+  const options = R.propOr({}, 'options', config);
+  const restConfig = R.omit('options', config);
+
+  if (R.has('options', config)) {
+    logger.warn('Deprecated usage of `options` in Dredd configuration.');
   }
 
-  applyLoggingOptions(outConfig.options);
+  return R.mergeDeepLeft(options, restConfig);
+}
 
-  // Log accumulated errors and warnings
-  coerceResult.errors.forEach(message => logger.error(message));
-  coerceResult.warnings.forEach(message => logger.warn(message));
+const defaultConfig = {
+  http: {},
+  server: null,
+  // emitter: new EventEmitter(), // todo Merge + preserve prototype?
+  custom: {
+    cwd: process.cwd(),
+  },
+  path: [],
+  apiDescriptions: [],
+  'dry-run': false,
+  reporter: null,
+  output: null,
+  header: null,
+  user: null,
+  'inline-errors': false,
+  details: false,
+  method: [],
+  only: [],
+  color: true,
+  loglevel: 'warn',
+  sorted: false,
+  names: false,
+  hookfiles: [],
+  language: 'nodejs',
+  'hooks-worker-timeout': 5000,
+  'hooks-worker-connect-timeout': 1500,
+  'hooks-worker-connect-retry': 500,
+  'hooks-worker-after-connect-wait': 100,
+  'hooks-worker-term-timeout': 5000,
+  'hooks-worker-term-retry': 500,
+  'hooks-worker-handler-host': '127.0.0.1',
+  'hooks-worker-handler-port': 61321,
+};
 
-  // Abort Dredd if there has been any errors
-  if (coerceResult.errors.length) {
+function resolveConfig(config) {
+  const inConfig = R.compose(
```

Comment:
I don't think we should call it `inConfig` anymore, it isn't the object on input of the `resolveConfig` function. Perhaps something along the lines of `configWithDefaults`?